In [1]:
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('telecom_churn.csv')

In [3]:
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null float64
Total eve minutes         3333 non-null float64
Total eve calls           3333 non-null int64
Total eve charge          3333 non-null float64
Total night minutes       3333 non-null float64
Total night calls         3333 non-null int64
Total night charge        3333 non-null float64
Total intl minutes        3333 non-null float64
Total intl calls          3333 non-null int64
Total intl charge         3333 non-null float64
Customer service calls    3333 non-null int64


In [5]:
df.shape

(3333, 20)

In [6]:
df['International plan'] = df['International plan'].map({'Yes': 1, 'No': 0})

In [7]:
df['Voice mail plan'] = df['Voice mail plan'].map({'Yes': 1, 'No': 0})

In [8]:
df['Churn'] = df['Churn'].astype('int')

In [9]:
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415                   0                1   
1    OH             107        415                   0                1   
2    NJ             137        415                   0                0   
3    OH              84        408                   1                0   
4    OK              75        415                   1                0   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

In [43]:
import numpy as np

In [44]:
y.value_counts(normalize='True') #85% change are not loyal, and 15% are loyal customers

0    0.855086
1    0.144914
Name: Churn, dtype: float64

### save states as a series and drop for now from the dataframe

In [10]:
states = df.pop('State')

### Split dataframe into a matrix X and vector y which form the train set 

In [11]:
X, y = df.drop('Churn', axis=1), df['Churn']

In [12]:
X.shape, y.shape

((3333, 18), (3333,))

In [13]:
from sklearn.model_selection import train_test_split

library 

In [14]:
X_train, X_holdout, y_train, y_holdout =  train_test_split(X, y, test_size=.3, random_state=17)

In [15]:
X_train.shape, X_holdout.shape

((2333, 18), (1000, 18))

In [16]:
from sklearn.tree import DecisionTreeClassifier

In [17]:
tree = DecisionTreeClassifier(random_state=17)

"fit-predict"

In [18]:
tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=17)

In [19]:
from sklearn.metrics import accuracy_score

In [20]:
pred_holdout = tree.predict(X_holdout)

In [21]:
pred_holdout.shape, y_holdout.shape

((1000,), (1000,))

In [22]:
accuracy_score(y_holdout, pred_holdout)

0.921

In [25]:
from sklearn.model_selection import GridSearchCV, StratifiedKFold

In [26]:
params = {'max_depth': np.arange(2,11), 'min_samples_leaf': np.arange(1,11)}

In [27]:
skf = StratifiedKFold(n_splits= 5, shuffle=True, random_state=17)

In [28]:
best_tree = GridSearchCV(estimator = tree, param_grid = params, cv = skf, n_jobs = -1,
                        verbose = 1)

In [29]:
best_tree.fit(X_train, y_train)

Fitting 5 folds for each of 90 candidates, totalling 450 fits


GridSearchCV(cv=StratifiedKFold(n_splits=5, random_state=17, shuffle=True),
             estimator=DecisionTreeClassifier(random_state=17), n_jobs=-1,
             param_grid={'max_depth': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10]),
                         'min_samples_leaf': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])},
             verbose=1)

In [30]:
best_tree.best_params_

{'max_depth': 6, 'min_samples_leaf': 4}

In [31]:
best_tree.best_estimator_

DecisionTreeClassifier(max_depth=6, min_samples_leaf=4, random_state=17)

**Cross-Validation assessment of model qualiy**

In [32]:
best_tree.best_score_

0.9438485079633493

**holdout assessment**

In [33]:
pred_holdout_better = best_tree.predict(X_holdout)

In [34]:
accuracy_score(y_holdout, pred_holdout_better)

0.94

In [35]:
from sklearn.tree import export_graphviz

In [36]:
export_graphviz(decision_tree=best_tree.best_estimator_,
               out_file='tree.dot', filled=True,
                feature_names=df.drop('Churn', axis=1).columns)

In [39]:
tree = DecisionTreeClassifier(random_state=17, max_depth=3).fit(X_train, y_train)

In [40]:
export_graphviz(decision_tree=tree,
               out_file='tree_depth3.dot', filled=True,
                feature_names=df.drop('Churn', axis=1).columns)

In [41]:
from io import StringIO
import pydotplus
from ipywidgets import Image

In [42]:
dot_data= StringIO()
export_graphviz(decision_tree=tree,
               out_file=dot_data, filled=True,
               feature_names=df.drop('Churn',axis=1).columns)
graph = pydotplus.graph_from_dot_data(dot_data.getvalue())
Image(value=graph.create_png())

Image(value=b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x04\xfd\x00\x00\x01\xf1\x08\x06\x00\x00\x00\x16Y\xc9…

**initially we had 2333 samples the distribution was 350 churned and others i.e 1983 are loyal customers and the more loyal customers we got
the brighter is going to be cell so orange one is going to be a for a good client and blue ones are going to be to show cells where have more 
bad clients/customers**